subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/Strings.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.instantexecution.serialization
+
+import org.gradle.internal.serialize.Decoder
+import org.gradle.internal.serialize.Encoder
+
+
+class WriteStrings {
+
+    fun writeString(string: CharSequence, encoder: Encoder) = encoder.run {
+        string.toString().let { string ->
+            val id = strings[string]
+            if (id != null) {
+                writeSmallInt(id)
+            } else {
+                val newId = strings.size
+                writeSmallInt(-1)
```

Comment:
I guess it would be slightly more efficient, if you use `0` as the value for "new string follows", then use 1-based indexes.

As far as I understand, `-1` takes more bytes on the wire than small positive numbers.

Have you considered it?

Reply:
I have not. Thanks for the suggestion, @vlsi, I'll give it a shot.

Reply:
It does generate smaller files and that seems to contribute to slightly quicker loading times:

![Screen Shot 2020-07-17 at 15 11 51](https://user-images.githubusercontent.com/51689/87818008-4dfb5300-c840-11ea-8fa9-088590f6a4ad.png)

File sizes, after and before:

```
.gradle/configuration-cache/5e1i6cnxrii7zkzisc803t7d1:
total 6712
drwx------  4 bamboo  staff      128 Jul 17 15:00 .
drwxr-xr-x  6 bamboo  staff      192 Jul 17 15:00 ..
-rw-------  1 bamboo  staff   207801 Jul 17 15:00 fingerprint.bin
-rw-------  1 bamboo  staff  2948875 Jul 17 15:00 state.bin

.gradle/configuration-cache/kgvavkfdjmbc5bbslvdh6z3r:
total 7104
drwx------  4 bamboo  staff      128 Jul 17 14:59 .
drwxr-xr-x  6 bamboo  staff      192 Jul 17 15:00 ..
-rw-------  1 bamboo  staff   212045 Jul 17 14:59 fingerprint.bin
-rw-------  1 bamboo  staff  2969466 Jul 17 14:59 state.bin
```